File: executor/agents/claude_code/multimodal_prompt.py
# SPDX-FileCopyrightText: 2025 Weibo, Inc.
#
# SPDX-License-Identifier: Apache-2.0

"""
Multimodal prompt utilities for Claude Code agent.

Handles conversion between OpenAI Responses API vision format and
Anthropic Messages API format, and creates async generators for
the Claude SDK's multimodal query path.
"""

import base64
import logging
import os
import re
import uuid
from datetime import datetime
from typing import Any, AsyncGenerator, Union

logger = logging.getLogger(__name__)


# MIME type to file extension mapping
_MIME_TO_EXT = {
    "image/png": ".png",
    "image/jpeg": ".jpg",
    "image/jpg": ".jpg",
    "image/gif": ".gif",
    "image/webp": ".webp",
    "image/bmp": ".bmp",
    "image/svg+xml": ".svg",
}


def is_vision_prompt(prompt: Union[str, list]) -> bool:
    """Check if a prompt contains vision content blocks.

    Args:
        prompt: Either a string prompt or a list of content blocks.

    Returns:
        True if prompt is a list containing vision content blocks.
    """
    if not isinstance(prompt, list) or len(prompt) == 0:
        return False
    return any(
        isinstance(block, dict) and block.get("type") in ("input_image", "image")
        for block in prompt
    )


def append_text_to_vision_prompt(
    prompt: list[dict[str, Any]],
    text: str,
    prepend: bool = False,
) -> list[dict[str, Any]]:
    """Append or prepend text to the text block in a vision content list.

    If an input_text block exists, the text is added to it.
    Otherwise, a new input_text block is inserted at the appropriate position.

    Args:
        prompt: List of OpenAI Responses API content blocks.
        text: Text to append or prepend.
        prepend: If True, prepend text; otherwise append.

    Returns:
        Updated list of content blocks (new list, original not mutated).
    """
    result = [block.copy() for block in prompt]

    # Find existing text block
    for block in result:
        if block.get("type") == "input_text":
            if prepend:
                block["text"] = text + "\n\n" + block["text"]
            else:
                block["text"] = block["text"] + "\n" + text
            return result

    # No text block found — create one
    new_block = {"type": "input_text", "text": text}
    if prepend:
        result.insert(0, new_block)
    else:
        result.append(new_block)
    return result


def convert_openai_to_anthropic_content(
    content_blocks: list[dict[str, Any]],
) -> list[dict[str, Any]]:
    """Convert OpenAI Responses API content blocks to Anthropic Messages API format.

    Conversions:
    - input_text -> text
    - input_image (data URI) -> image with base64 source

    Args:
        content_blocks: List of OpenAI Responses API content blocks.

    Returns:
        List of Anthropic Messages API content blocks.
    """
    anthropic_blocks: list[dict[str, Any]] = []

    for block in content_blocks:
        block_type = block.get("type", "")

        if block_type == "input_text":
            anthropic_blocks.append(
                {
                    "type": "text",
                    "text": block.get("text", ""),
                }
            )

        elif block_type == "input_image":
            image_url = block.get("image_url", "")
            media_type, data = _parse_data_uri(image_url)
            anthropic_blocks.append(
                {
                    "type": "image",
                    "source": {
                        "type": "base64",
                        "media_type": media_type,
                        "data": data,
                    },
                }
            )

        else:
            # Pass through unknown block types as-is
            anthropic_blocks.append(block)

    return anthropic_blocks


async def create_multimodal_query(
    anthropic_content: list[dict[str, Any]],
) -> AsyncGenerator[dict[str, Any], None]:
    """Create an async generator that yields a multimodal user message for the SDK.

    The Claude SDK query() method accepts either a string or an AsyncIterable
    of message dicts. This function creates the async iterable path for
    multimodal content.

    Args:
        anthropic_content: Anthropic Messages API content blocks.

    Yields:
        A single user message dict compatible with the Claude SDK.
    """
    yield {
        "type": "user",
        "message": {
            "role": "user",
            "content": anthropic_content,
        },
    }


def save_vision_images(
    prompt: list[dict[str, Any]],
    task_id: str = "",
) -> list[str]:
    """Save base64 images from vision prompt to disk.

    Images are saved to ~/.wegent-executor/docs/pics/<task_id>/.
    The WEGENT_EXECUTOR_HOME env var is respected.

    Args:
        prompt: List of OpenAI Responses API content blocks.
        task_id: Task ID for organizing images into subdirectories.

    Returns:
        List of saved file paths.
    """
    from executor.config.config import WEGENT_EXECUTOR_HOME

    month_dir = datetime.now().strftime("%Y%m")
    pics_dir = os.path.join(WEGENT_EXECUTOR_HOME, "docs", "pics", month_dir)
    os.makedirs(pics_dir, exist_ok=True)

    saved_paths: list[str] = []
    for block in prompt:
        if block.get("type") != "input_image":
            continue

        image_url = block.get("image_url", "")
        media_type, data = _parse_data_uri(image_url)
        ext = _MIME_TO_EXT.get(media_type, ".png")
        prefix = f"{task_id}_" if task_id else ""
        filename = f"{prefix}{uuid.uuid4().hex[:12]}{ext}"
        filepath = os.path.join(pics_dir, filename)

        try:
            image_bytes = base64.b64decode(data)
            with open(filepath, "wb") as f:
                f.write(image_bytes)
            saved_paths.append(filepath)
            logger.info(
                "[multimodal] Saved image: %s (%d bytes)", filepath, len(image_bytes)
            )
        except Exception as e:
            logger.error("[multimodal] Failed to save image: %s", e)

    return saved_paths


def _parse_data_uri(data_uri: str) -> tuple[str, str]:
    """Parse a data URI into media type and base64 data.

    Args:
        data_uri: A data URI string like "data:image/png;base64,iVBOR..."

    Returns:
        Tuple of (media_type, base64_data). Defaults to
        ("image/png", data_uri) if parsing fails.
    """
    match = re.match(r"^data:([^;]+);base64,(.+)$", data_uri, re.DOTALL)
    if match:
        return match.group(1), match.group(2)
    return "image/png", data_uri

File: executor/src/attachments/multimodal_prompt.rs
// SPDX-FileCopyrightText: 2026 Weibo, Inc.
//
// SPDX-License-Identifier: Apache-2.0

use std::{
    fs, io,
    path::{Path, PathBuf},
    sync::atomic::{AtomicU64, Ordering},
    time::{SystemTime, UNIX_EPOCH},
};

use base64::{engine::general_purpose, Engine as _};
use serde_json::{json, Value};

use crate::image_preprocessor::prepare_image_bytes_for_model;

static SAVE_COUNTER: AtomicU64 = AtomicU64::new(0);

pub fn is_vision_prompt(prompt: &Value) -> bool {
    prompt.as_array().into_iter().flatten().any(|block| {
        matches!(
            block.get("type").and_then(Value::as_str),
            Some("input_image" | "image" | "localImage" | "local_image")
        )
    })
}

pub fn append_text_to_vision_prompt(prompt: &Value, text: &str, prepend: bool) -> Value {
    let Some(blocks) = prompt.as_array() else {
        return prompt.clone();
    };

    let mut result = blocks.clone();
    for block in &mut result {
        if block.get("type").and_then(Value::as_str) != Some("input_text") {
            continue;
        }

        let existing = block.get("text").and_then(Value::as_str).unwrap_or("");
        let updated_text = if prepend {
            format!("{text}\n\n{existing}")
        } else {
            format!("{existing}\n{text}")
        };
        if let Some(object) = block.as_object_mut() {
            object.insert("text".to_owned(), Value::String(updated_text));
        }
        return Value::Array(result);
    }

    let text_block = json!({"type": "input_text", "text": text});
    if prepend {
        result.insert(0, text_block);
    } else {
        result.push(text_block);
    }
    Value::Array(result)
}

pub fn convert_openai_to_anthropic_content(content_blocks: &Value) -> Value {
    let Some(blocks) = content_blocks.as_array() else {
        return content_blocks.clone();
    };

    Value::Array(blocks.iter().filter_map(convert_content_block).collect())
}

pub async fn convert_openai_to_anthropic_content_async(content_blocks: Value) -> Value {
    tokio::task::spawn_blocking(move || convert_openai_to_anthropic_content(&content_blocks))
        .await
        .expect("multimodal conversion task panicked")
}

pub fn create_multimodal_query(anthropic_content: &Value) -> Value {
    json!([{
        "type": "user",
        "message": {
            "role": "user",
            "content": anthropic_content,
        },
    }])
}

pub fn save_vision_images(prompt: &Value, task_id: Option<&str>) -> io::Result<Vec<PathBuf>> {
    save_vision_images_to(prompt, &default_executor_home(), task_id)
}

pub fn save_vision_images_to(
    prompt: &Value,
    executor_home: &Path,
    task_id: Option<&str>,
) -> io::Result<Vec<PathBuf>> {
    let Some(blocks) = prompt.as_array() else {
        return Ok(Vec::new());
    };

    let pics_dir = executor_home
        .join("docs")
        .join("pics")
        .join(current_month_dir());
    fs::create_dir_all(&pics_dir)?;

    let mut saved_paths = Vec::new();
    for block in blocks {
        if block.get("type").and_then(Value::as_str) != Some("input_image") {
            continue;
        }

        let image_url = block
            .get("image_url")
            .or_else(|| block.get("url"))
            .and_then(Value::as_str)
            .unwrap_or("");
        let (media_type, data) = parse_data_uri(image_url);
        let Ok(image_bytes) = general_purpose::STANDARD.decode(data) else {
            continue;
        };

        let prefix = task_id
            .filter(|value| !value.is_empty())
            .map(|value| format!("{value}_"))
            .unwrap_or_default();
        let path = pics_dir.join(format!(
            "{prefix}{}{}",
            unique_filename_stem(),
            image_extension(&media_type)
        ));
        fs::write(&path, image_bytes)?;
        saved_paths.push(path);
    }

    Ok(saved_paths)
}

pub fn parse_data_uri(data_uri: &str) -> (String, String) {
    let Some(rest) = data_uri.strip_prefix("data:") else {
        return ("image/png".to_owned(), data_uri.to_owned());
    };
    let Some((media_type, data)) = rest.split_once(";base64,") else {
        return ("image/png".to_owned(), data_uri.to_owned());
    };
    if media_type.is_empty() || data.is_empty() {
        return ("image/png".to_owned(), data_uri.to_owned());
    }
    (media_type.to_owned(), data.to_owned())
}

fn convert_content_block(block: &Value) -> Option<Value> {
    match block.get("type").and_then(Value::as_str).unwrap_or("") {
        "input_text" => text_value(block)
            .filter(|text| !text.trim().is_empty())
            .map(|text| json!({"type": "text", "text": text})),
        "text" => text_value(block)
            .filter(|text| !text.trim().is_empty())
            .map(|_| block.clone()),
        "input_image" => {
            let image_url = block
                .get("image_url")
                .or_else(|| block.get("url"))
                .and_then(Value::as_str)
                .unwrap_or("");
            let (media_type, data) = parse_data_uri(image_url);
            let (media_type, data) = prepare_base64_image_for_model(&media_type, &data);
            Some(json!({
                "type": "image",
                "source": {
                    "type": "base64",
                    "media_type": media_type,
                    "data": data,
                },
            }))
        }
        _ => Some(block.clone()),
    }
}

fn text_value(block: &Value) -> Option<&str> {
    block.get("text").and_then(Value::as_str)
}

fn prepare_base64_image_for_model(media_type: &str, data: &str) -> (String, String) {
    let Ok(image_data) = general_purpose::STANDARD.decode(data) else {
        return (media_type.to_owned(), data.to_owned());
    };

    let prepared = prepare_image_bytes_for_model(&image_data, media_type, None);
    if !prepared.resized {
        return (media_type.to_owned(), data.to_owned());
    }

    (
        prepared.mime_type,
        general_purpose::STANDARD.encode(prepared.data),
    )
}

fn default_executor_home() -> PathBuf {
    if let Some(home) = std::env::var_os("WEGENT_EXECUTOR_HOME") {
        return PathBuf::from(home);
    }
    dirs::home_dir()
        .map(|home| home.join(".wegent-executor"))
        .unwrap_or_else(|| PathBuf::from(".wegent-executor"))
}

fn image_extension(media_type: &str) -> &'static str {
    match media_type.to_ascii_lowercase().as_str() {
        "image/jpeg" | "image/jpg" => ".jpg",
        "image/gif" => ".gif",
        "image/webp" => ".webp",
        "image/bmp" => ".bmp",
        "image/svg+xml" => ".svg",
        _ => ".png",
    }
}

fn unique_filename_stem() -> String {
    let counter = SAVE_COUNTER.fetch_add(1, Ordering::Relaxed);
    let nanos = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map(|duration| duration.as_nanos())
        .unwrap_or_default();
    format!("{:x}{nanos:x}{counter:x}", std::process::id())
}

fn current_month_dir() -> String {
    let days_since_epoch = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map(|duration| duration.as_secs() / 86_400)
        .unwrap_or_default() as i64;
    let (year, month) = year_month_from_unix_days(days_since_epoch);
    format!("{year:04}{month:02}")
}

fn year_month_from_unix_days(days_since_epoch: i64) -> (i32, u32) {
    let days = days_since_epoch + 719_468;
    let era = if days >= 0 { days } else { days - 146_096 } / 146_097;
    let day_of_era = days - era * 146_097;
    let year_of_era =
        (day_of_era - day_of_era / 1_460 + day_of_era / 36_524 - day_of_era / 146_096) / 365;
    let day_of_year = day_of_era - (365 * year_of_era + year_of_era / 4 - year_of_era / 100);
    let month_parameter = (5 * day_of_year + 2) / 153;
    let month = month_parameter + if month_parameter < 10 { 3 } else { -9 };
    let year = year_of_era + era * 400 + if month <= 2 { 1 } else { 0 };
    (year as i32, month as u32)
}
